### PR TITLE
Adaptor signatures for Scala.js

### PR DESCRIPTION
**Pull request by rorp**

### CODE REVIEW

**Review by benthecarman — APPROVED** [review]
Looks correct to me, however, I am not very familiar with the spec for this

File: crypto/.js/src/main/scala/org/bitcoins/crypto/BCryptoCryptoRuntime.scala [hunk 1]
```diff
@@ -109,11 +109,17 @@ trait BCryptoCryptoRuntime extends CryptoRuntime {
       signature: ECDigitalSignature,
       message: ByteVector): (ECPublicKey, ECPublicKey) = {
     val keyBytes =
-      ecdsa.recover(message, signature.bytes, param = 0, compress = true)
+      if (signature.isDEREncoded)
+        ecdsa.recoverDER(message, signature.bytes, param = 0, compress = true)
```
Comment (Christewart) [inline crypto/.js/src/main/scala/org/bitcoins/crypto/BCryptoCryptoRuntime.scala:112]: what is `param= 0` for ?

File: crypto/src/main/scala/org/bitcoins/crypto/AdaptorUtil.scala [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+  // Compute s' = k^-1 * (dataToSign + rx*privateKey)
+  private def adaptorSignHelper(
+      dataToSign: ByteVector,
+      k: FieldElement,
+      r: ECPublicKey,
+      privateKey: ECPrivateKey): FieldElement = {
+    CryptoUtil.decodePoint(r) match {
+      case ECPointInfinity =>
+        throw new IllegalArgumentException("Invalid point")
```
Comment (Christewart) [inline crypto/src/main/scala/org/bitcoins/crypto/AdaptorUtil.scala]: nit: 

```scala
throw new IllegalArgumentException(s"Invalid point, got ${ECPointInfinity}")
```

Generally putting the input in the error message is _really_ helpful when debugging.

File: crypto/src/main/scala/org/bitcoins/crypto/CryptoRuntime.scala [hunk 3]
```diff
@@ -134,6 +134,8 @@ trait CryptoRuntime {
 
   def publicKey(privateKey: ECPrivateKey): ECPublicKey
 
+  def publicKeyConvert(key: ECPublicKey, compressed: Boolean): ECPublicKey
```
Comment (Christewart) [inline crypto/src/main/scala/org/bitcoins/crypto/CryptoRuntime.scala:140]: Scaladoc please, what is it converting to?

If I understand what this method is meant to do, i would recommend removing the `compressed` parameter and calling it 

```scala
def toCompressedPublicKey(key: ECPublicKey): ECPublicKey
``` 

?

File: crypto/src/main/scala/org/bitcoins/crypto/DERSignatureUtil.scala [hunk 4]
```diff
@@ -37,12 +37,16 @@ sealed abstract class DERSignatureUtil {
       val thirdByteIs0x02 = bytes(2) == 0x02
       //this is the size of the r value in the signature
       val rSize = bytes(3)
-
-      //this 0x02 separates the r and s value )in the signature
-      val second0x02Exists = bytes(rSize + 4) == 0x02
-
-      firstByteIs0x30 && signatureLengthIsCorrect && thirdByteIs0x02 &&
-      second0x02Exists
+      val sOffset = rSize + 4
```
Comment (Christewart) [inline crypto/src/main/scala/org/bitcoins/crypto/DERSignatureUtil.scala:40]: What is the bug here? Just an `OutOfBoundsException` if an invalid DER encoded `ByteVector` is passed into this function?

**Review by Christewart — COMMENTED** [review]
Also looking very good! Mostly nits

File: crypto/.js/src/main/scala/org/bitcoins/crypto/ECDSA.scala [hunk 5]
```diff
@@ -41,11 +41,21 @@ class ECDSA(
...
       msg: Buffer,
       sig: Buffer,
       param: Byte,
       compress: Boolean): Buffer = js.native
 
+  def recoverDER(
+      msg: Buffer,
+      sig: Buffer,
+      param: Byte,
+      compress: Boolean): Buffer = js.native
```
Comment (nkohen) [inline crypto/.js/src/main/scala/org/bitcoins/crypto/ECDSA.scala:58]: Seeing as DER is the default in Bitcoin, I find it concerning to have `recover` and `verify` and `sign` not be doing DER-related things?

**Pull request merged**